include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -482,41 +492,487 @@ namespace internal
+      std::vector<std::vector<unsigned int>> cell_local_children_indices;
+      std::vector<unsigned int>              lexicographic_numbering;
+
+      {
+        const auto &fe             = dof_handler_fine.get_fe();
+        const auto &reference_cell = fe.reference_cell();
+
+        const bool is_feq =
+          fe.n_base_elements() == 1 &&
+          (dynamic_cast<const FE_Q<dim> *>(&fe.base_element(0)) != nullptr);
+
+        n_dofs_per_cell_coarse = fe.n_dofs_per_cell();
+        n_dofs_per_cell_fine =
+          is_feq ?
+            (fe.n_components() * Utilities::pow(2 * fe.degree + 1, dim)) :
+            (fe.n_dofs_per_cell() * GeometryInfo<dim>::max_children_per_cell);
+
+        cell_local_children_indices =
+          get_child_offsets<dim>(n_dofs_per_cell_coarse,
+                                 is_feq ? fe.degree : (fe.degree + 1),
+                                 fe.degree);
+
+        if (reference_cell.is_hyper_cube())
+          {
+            const Quadrature<1> dummy_quadrature(
+              std::vector<Point<1>>(1, Point<1>()));
+            internal::MatrixFreeFunctions::ShapeInfo<double> shape_info;
+            shape_info.reinit(dummy_quadrature, fe, 0);
+            lexicographic_numbering = shape_info.lexicographic_numbering;
+          }
+        else
+          {
+            const auto dummy_quadrature =
+              reference_cell.template get_gauss_type_quadrature<dim>(1);
+            internal::MatrixFreeFunctions::ShapeInfo<double> shape_info;
+            shape_info.reinit(dummy_quadrature, fe, 0);
+            lexicographic_numbering = shape_info.lexicographic_numbering;
+          }
+      }
+
+      if (this->mg_level_fine == numbers::invalid_unsigned_int)
+        {
+          is_locally_owned_dofs = dof_handler_fine.locally_owned_dofs();
+
+          is_locally_relevant_dofs.set_size(is_locally_owned_dofs.size());
+
+          std::vector<types::global_dof_index> dof_indices_cell;
+          std::vector<types::global_dof_index> dof_indices_patch;
+
+          loop_over_active_or_level_cells(
+            dof_handler_coarse,
+            numbers::invalid_unsigned_int,
+            [&](const auto &cell) {
+              const auto cell_id = cell->id();
+
+              const auto cell_fine_raw =
+                dof_handler_fine.get_triangulation().create_cell_iterator(
+                  cell_id);
+
+              if (cell_fine_raw->has_children() == false)
+                {
+                  const auto cell_fine =
+                    cell_fine_raw->as_dof_handler_iterator(dof_handler_fine);
+
+                  dof_indices_cell.resize(
+                    cell_fine->get_fe().n_dofs_per_cell());
+                  cell_fine->get_dof_indices(dof_indices_cell);
+                  indices.insert(indices.end(),
+                                 dof_indices_cell.begin(),
+                                 dof_indices_cell.end());
```

Comment:
Since I only see the use for the relevant dofs, wouldn't it make the code faster to only insert the entries that not in the `is_locally_owned_dofs`?

Reply:
Is the missing `if` statement here and the equivalent code a few lines below, compared to the old state, the reason for the high cost in `std::sort` that you told me when we talked in person?

Reply:
I didn't think about the check with `is_locally_owned_dofs`. I will try this out!

Reply:
I went with the suggestion. It was a bit faster than what I have implemented :+1: